## Instructions

1. **Labeling & Peer Grading:** Your homework will be peer graded. To stay anonymous, avoid using your name and label your file with the last four digits of your student ID (e.g., HW2_Solutions_3938).

2. **Submission:** Submit both your IPython notebook (.ipynb) and an HTML file of the notebook to Canvas under Assignments → HW 2 → Submit Assignment. After submitting, download and check the files to make sure that you've uploaded the correct versions. Both files are required for your HW to be graded.
3. 
 <font color='red'> No pdf file required so write all the details in your ipynb file.</font>
2. **AI Use Policy:** Solve each problem independently by yourself. Use AI tools like ChatGPT or Google Gemini for brainstorming and learning only—copying AI-generated content is prohibited. You do not neeViolations will lead to penalties, up to failing the course.

3. **Problem Structure:** <font color='red'>Break down each problem ( already done in most problems) into three interconnected parts and implement each in separate code cells. Ensure that each part logically builds on the previous one. Include comments in your code to explain its purpose, followed by a Markdown cell analyzing what was achieved. After completing all parts, add a final Markdown cell reflecting on your overall approach, discussing any challenges faced, and explaining how you utilized AI tools in your process.
</font>
4. **Deadlines & Academic Integrity:** This homework is due on 10/01/2024 at midnight. <font color='red'>Disclosure of this assignment and assignment answers to anyone or any website is a contributory infringement of academic dishonesty at ISU. Do not share or post course materials without the express written consent of the copyright holder and instructor. The class will follow Iowa State University’s policy on academic dishonesty. Anyone suspected of academic dishonesty will be reported to the Dean of Students Office.</font>

#### Each problem is worth 20 points. Total $\bf 20\times 5 = 100$.

### Problem 1. 
Upload the newimdb data and read the details on the columns about the data following the link https://www.imdb.com/interfaces/.
* Begin the data cleaning process by addressing duplicate entries in the dataset, with a particular focus on the 'category' and 'directors' columns. Remove any redundant rows and determine the best approach for handling variations within these columns. You may choose to either aggregate the diverse values into lists or select the most frequent entry for each instance; explain the reasoning behind your chosen method. Subsequently, identify and eliminate irrelevant columns, such as 'Unnamed: 0', which do not contribute meaningful information for regression or classification tasks. Provide a clear justification for the removal of each column, taking into account their potential significance in model development. Your explanation should demonstrate a thorough understanding of the dataset's structure and the relevance of each feature to the intended analysis.
* Analyze the dataset for missing values, particularly in columns like 'runtimeMinutes' and 'genres'. Determine whether to remove rows with missing data or impute values using appropriate statistical methods (mean, median, or mode), and justify your approach. For feature engineering, transform non-numerical columns such as 'genres' and 'directors' using one-hot encoding or label encoding techniques. Consider grouping genres into broader categories to potentially improve model performance. Provide clear explanations for your strategies in handling missing values and implementing feature engineering, ensuring your decisions are data-driven and aligned with the goals of the analysis.
  
* Conduct exploratory data analysis (EDA) on all remaining columns. Handle non-numerical values appropriately, convert data types as needed, and apply transformations or standardization where necessary. Implement dummy coding for categorical variables. For classification purposes, create a new 'Rating' column derived from 'averageRating'. Instead of using a predetermined scale, devise a binning strategy that results in a balanced class distribution across 5 rating categories (1 to 5 stars). Explain your chosen binning method and analyze the resulting distribution. Discuss how this balanced approach may impact your classification model's performance compared to using the original 'averageRating' values.

## Libraries

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)

## Config

In [2]:
data_loc = "/Users/gabrielvictorgomesferreira/Library/Mobile Documents/com~apple~CloudDocs/Work/ISU Classes/MIS 546 - Advanced Business Analytics/Data/"
file_name = "newimdb.csv"

## Import dataset

In [3]:
imdb_df = pd.read_csv(data_loc + file_name)
print("Data dimensions: ", imdb_df.shape)
imdb_df.head(20)

Data dimensions:  (2795295, 14)


Unnamed: 0     tconst        types  isOriginalTitle titleType  \
0      6475881  tt0111596  imdbDisplay              0.0     movie   
1      6475882  tt0111596  imdbDisplay              0.0     movie   
2      6475884  tt0111596  imdbDisplay              0.0     movie   
3      6475886  tt0111596  imdbDisplay              0.0     movie   
4      6475887  tt0111596  imdbDisplay              0.0     movie   
5      6475889  tt0111596  imdbDisplay              0.0     movie   
6      6475890  tt0111596  imdbDisplay              0.0     movie   
7      6475891  tt0111596     original              1.0     movie   
8      6475892  tt0111596     original              1.0     movie   
9      6475894  tt0111596     original              1.0     movie   
10     6475896  tt0111596     original              1.0     movie   
11     6475897  tt0111596     original              1.0     movie   
12     6475899  tt0111596     original              1.0     movie   
13     6475900  tt0111596     original              1.0     movie   
14     6955478  tt0118589  imdbDisplay              0.0     movie   
15     6955479  tt0118589  imdbDisplay              0.0     movie   
16     6955480  tt0118589  imdbDisplay              0.0     movie   
17     6955483  tt0118589  imdbDisplay              0.0     movie   
18     6955484  tt0118589  imdbDisplay              0.0     movie   
19     6955486  tt0118589  imdbDisplay              0.0     movie   

   originalTitle  isAdult  startYear runtimeMinutes               genres  \
0          Vazir      0.0       2015             \N   Action,Crime,Drama   
1          Vazir      0.0       2015             \N   Action,Crime,Drama   
2          Vazir      0.0       2015             \N   Action,Crime,Drama   
3          Vazir      0.0       2015             \N   Action,Crime,Drama   
4          Vazir      0.0       2015             \N   Action,Crime,Drama   
5          Vazir      0.0       2015             \N   Action,Crime,Drama   
6          Vazir      0.0       2015             \N   Action,Crime,Drama   
7          Vazir      0.0       2015             \N   Action,Crime,Drama   
8          Vazir      0.0       2015             \N   Action,Crime,Drama   
9          Vazir      0.0       2015             \N   Action,Crime,Drama   
10         Vazir      0.0       2015             \N   Action,Crime,Drama   
11         Vazir      0.0       2015             \N   Action,Crime,Drama   
12         Vazir      0.0       2015             \N   Action,Crime,Drama   
13         Vazir      0.0       2015             \N   Action,Crime,Drama   
14       Glitter      0.0       2001            104  Drama,Music,Romance   
15       Glitter      0.0       2001            104  Drama,Music,Romance   
16       Glitter      0.0       2001            104  Drama,Music,Romance   
17       Glitter      0.0       2001            104  Drama,Music,Romance   
18       Glitter      0.0       2001            104  Drama,Music,Romance   
19       Glitter      0.0       2001            104  Drama,Music,Romance   

     directors         category  averageRating  numVotes  
0   nm13648797           editor            8.1      63.0  
1   nm13648797            actor            8.1      63.0  
2   nm13648797          actress            8.1      63.0  
3   nm13648797         director            8.1      63.0  
4   nm13648797           writer            8.1      63.0  
5   nm13648797         composer            8.1      63.0  
6   nm13648797  cinematographer            8.1      63.0  
7   nm13648797           editor            8.1      63.0  
8   nm13648797            actor            8.1      63.0  
9   nm13648797          actress            8.1      63.0  
10  nm13648797         director            8.1      63.0  
11  nm13648797           writer            8.1      63.0  
12  nm13648797         composer            8.1      63.0  
13  nm13648797  cinematographer            8.1      63.0  
14   nm0193554  cinematographer            2.4   23686.0  
15   nm019355

## Data Exploration and Cleaning

In [4]:
imdb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2795295 entries, 0 to 2795294
Data columns (total 14 columns):
 #   Column           Dtype  
---  ------           -----  
 0   Unnamed: 0       int64  
 1   tconst           object 
 2   types            object 
 3   isOriginalTitle  float64
 4   titleType        object 
 5   originalTitle    object 
 6   isAdult          float64
 7   startYear        int64  
 8   runtimeMinutes   object 
 9   genres           object 
 10  directors        object 
 11  category         object 
 12  averageRating    float64
 13  numVotes         float64
dtypes: float64(4), int64(2), object(8)
memory usage: 298.6+ MB


### Firt Movie Sample: 'Vazir'

In [5]:
imdb_df[imdb_df['originalTitle'] == 'Vazir']

Unnamed: 0     tconst        types  isOriginalTitle titleType  \
0      6475881  tt0111596  imdbDisplay              0.0     movie   
1      6475882  tt0111596  imdbDisplay              0.0     movie   
2      6475884  tt0111596  imdbDisplay              0.0     movie   
3      6475886  tt0111596  imdbDisplay              0.0     movie   
4      6475887  tt0111596  imdbDisplay              0.0     movie   
5      6475889  tt0111596  imdbDisplay              0.0     movie   
6      6475890  tt0111596  imdbDisplay              0.0     movie   
7      6475891  tt0111596     original              1.0     movie   
8      6475892  tt0111596     original              1.0     movie   
9      6475894  tt0111596     original              1.0     movie   
10     6475896  tt0111596     original              1.0     movie   
11     6475897  tt0111596     original              1.0     movie   
12     6475899  tt0111596     original              1.0     movie   
13     6475900  tt0111596     original              1.0     movie   

   originalTitle  isAdult  startYear runtimeMinutes              genres  \
0          Vazir      0.0       2015             \N  Action,Crime,Drama   
1          Vazir      0.0       2015             \N  Action,Crime,Drama   
2          Vazir      0.0       2015             \N  Action,Crime,Drama   
3          Vazir      0.0       2015             \N  Action,Crime,Drama   
4          Vazir      0.0       2015             \N  Action,Crime,Drama   
5          Vazir      0.0       2015             \N  Action,Crime,Drama   
6          Vazir      0.0       2015             \N  Action,Crime,Drama   
7          Vazir      0.0       2015             \N  Action,Crime,Drama   
8          Vazir      0.0       2015             \N  Action,Crime,Drama   
9          Vazir      0.0       2015             \N  Action,Crime,Drama   
10         Vazir      0.0       2015             \N  Action,Crime,Drama   
11         Vazir      0.0       2015             \N  Action,Crime,Drama   
12         Vazir      0.0       2015             \N  Action,Crime,Drama   
13         Vazir      0.0       2015             \N  Action,Crime,Drama   

     directors         category  averageRating  numVotes  
0   nm13648797           editor            8.1      63.0  
1   nm13648797            actor            8.1      63.0  
2   nm13648797          actress            8.1      63.0  
3   nm13648797         director            8.1      63.0  
4   nm13648797           writer            8.1      63.0  
5   nm13648797         composer            8.1      63.0  
6   nm13648797  cinematographer            8.1      63.0  
7   nm13648797           editor            8.1      63.0  
8   nm13648797            actor            8.1      63.0  
9   nm13648797          actress            8.1      63.0  
10  nm13648797         director            8.1      63.0  
11  nm13648797           writer            8.1      63.0  
12  nm13648797         composer            8.1      63.0  
13  nm13648797  cinematographer            8.1      63.0

### Obs.:
In the above sample, all observations point to the same movie, but there are different title types in `types` column that only impact the binary `isOriginalTitle` column, which indicates whether the title is original. This variation doesn't seem to affect any other features. I will further explore this hypothesis using additional examples.

In [6]:
imdb_df['types'].value_counts()

types
original                1080513
imdbDisplay             1055184
\N                       395137
alternative               85040
working                   73868
festival                  42436
dvd                       36697
tv                        21941
video                      3283
dvdimdbDisplay             823
festivalimdbDisplay        142
imdbDisplaytv              103
imdbDisplayworking          46
imdbDisplayvideo            26
alternativedvd              20
alternativetv               13
alternativeworking          13
videoworking                 8
alternativefestival          2
Name: count, dtype: int64

In [7]:
imdb_df['titleType'].value_counts()

titleType
movie           1549698
short            659811
tvMovie          195005
tvSeries         187470
video            112483
videoGame         33181
tvMiniSeries      27829
tvSpecial         23492
tvShort            5360
tvEpisode           966
Name: count, dtype: int64

In [8]:
imdb_df[imdb_df['types'] == 'alternative']

Unnamed: 0     tconst        types  isOriginalTitle titleType  \
60          6961780  tt0118652  alternative              0.0     movie   
61          6961781  tt0118652  alternative              0.0     movie   
62          6961784  tt0118652  alternative              0.0     movie   
63          6961785  tt0118652  alternative              0.0     movie   
64          6961786  tt0118652  alternative              0.0     movie   
...             ...        ...          ...              ...       ...   
2795262    32998206  tt9916362  alternative              0.0     movie   
2795263    32998207  tt9916362  alternative              0.0     movie   
2795264    32998210  tt9916362  alternative              0.0     movie   
2795265    32998211  tt9916362  alternative              0.0     movie   
2795266    32998212  tt9916362  alternative              0.0     movie   

                 originalTitle  isAdult  startYear runtimeMinutes  \
60       The Attic Expeditions      0.0       2001            100   
61       The Attic Expeditions      0.0       2001            100   
62       The Attic Expeditions      0.0       2001            100   
63       The Attic Expeditions      0.0       2001            100   
64       The Attic Expeditions      0.0       2001            100   
...                        ...      ...        ...            ...   
2795262               Akelarre      0.0       2020             92   
2795263               Akelarre      0.0       2020             92   
2795264               Akelarre      0.0       2020             92   
2795265               Akelarre      0.0       2020             92   
2795266               Akelarre      0.0       2020             92   

                        genres  directors  category  averageRating  numVotes  
60       Comedy,Horror,Mystery  nm0440948  composer            4.9    1850.0  
61       Comedy,Horror,Mystery  nm0440948     actor            4.9    1850.0  
62       Comedy,Horror,Mystery  nm0440948   actress            4.9    1850.0  
63       Comedy,Horror,Mystery  nm0440948  director            4.9    1850.0  
64       Comedy,Horror,Mystery  nm0440948    writer            4.9    1850.0  
...                        ...        ...       ...            ...       ...  
2795262          Drama,History  nm1893148   actress            6.4    5179.0  
2795263          Drama,History  nm1893148     actor            6.4    5179.0  
2795264          Drama,History  nm1893148  director            6.4    5179.0  
2795265          Drama,History  nm1893148    writer            6.4    5179.0  
2795266          Drama,History  nm1893148  producer            6.4    5179.0  

[85040 rows x 14 columns]

In [9]:
imdb_df.iloc[0,8]

'\\N'

In [10]:
imdb_df[imdb_df['types'] == '\\N']

Unnamed: 0     tconst types  isOriginalTitle titleType  \
21          6955488  tt0118589    \N              0.0     movie   
22          6955489  tt0118589    \N              0.0     movie   
23          6955490  tt0118589    \N              0.0     movie   
24          6955493  tt0118589    \N              0.0     movie   
25          6955494  tt0118589    \N              0.0     movie   
...             ...        ...   ...              ...       ...   
2795280    32998342  tt9916538    \N              0.0     movie   
2795281    32998344  tt9916538    \N              0.0     movie   
2795282    32998346  tt9916538    \N              0.0     movie   
2795283    32998347  tt9916538    \N              0.0     movie   
2795284    32998350  tt9916538    \N              0.0     movie   

               originalTitle  isAdult  startYear runtimeMinutes  \
21                   Glitter      0.0       2001            104   
22                   Glitter      0.0       2001            104   
23                   Glitter      0.0       2001            104   
24                   Glitter      0.0       2001            104   
25                   Glitter      0.0       2001            104   
...                      ...      ...        ...            ...   
2795280  Kuambil Lagi Hatiku      0.0       2019            123   
2795281  Kuambil Lagi Hatiku      0.0       2019            123   
2795282  Kuambil Lagi Hatiku      0.0       2019            123   
2795283  Kuambil Lagi Hatiku      0.0       2019            123   
2795284  Kuambil Lagi Hatiku      0.0       2019            123   

                      genres  directors         category  averageRating  \
21       Drama,Music,Romance  nm0193554  cinematographer            2.4   
22       Drama,Music,Romance  nm0193554          actress            2.4   
23       Drama,Music,Romance  nm0193554            actor            2.4   
24       Drama,Music,Romance  nm0193554         director            2.4   
25       Drama,Music,Romance  nm0193554           writer            2.4   
...                      ...        ...              ...            ...   
2795280                Drama  nm4457074          actress            8.3   
2795281                Drama  nm4457074            actor            8.3   
2795282                Drama  nm4457074         director            8.3   
2795283                Drama  nm4457074           writer            8.3   
2795284                Drama  nm4457074         producer            8.3   

         numVotes  
21        23686.0  
22        23686.0  
23        23686.0  
24        23686.0  
25        23686.0  
...           ...  
2795280       6.0  
2795281       6.0  
2795282       6.0  
2795283       6.0  
2795284       6.0  

[395137 rows x 14 columns]

In [11]:
imdb_df[imdb_df['originalTitle'] == 'The Attic Expeditions']

Unnamed: 0     tconst        types  isOriginalTitle titleType  \
42     6961650  tt0118652  imdbDisplay              0.0     movie   
43     6961651  tt0118652  imdbDisplay              0.0     movie   
44     6961654  tt0118652  imdbDisplay              0.0     movie   
45     6961655  tt0118652  imdbDisplay              0.0     movie   
46     6961656  tt0118652  imdbDisplay              0.0     movie   
47     6961657  tt0118652  imdbDisplay              0.0     movie   
48     6961680  tt0118652     original              1.0     movie   
49     6961681  tt0118652     original              1.0     movie   
50     6961684  tt0118652     original              1.0     movie   
51     6961685  tt0118652     original              1.0     movie   
52     6961686  tt0118652     original              1.0     movie   
53     6961687  tt0118652     original              1.0     movie   
54     6961740  tt0118652          dvd              0.0     movie   
55     6961741  tt0118652          dvd              0.0     movie   
56     6961744  tt0118652          dvd              0.0     movie   
57     6961745  tt0118652          dvd              0.0     movie   
58     6961746  tt0118652          dvd              0.0     movie   
59     6961747  tt0118652          dvd              0.0     movie   
60     6961780  tt0118652  alternative              0.0     movie   
61     6961781  tt0118652  alternative              0.0     movie   
62     6961784  tt0118652  alternative              0.0     movie   
63     6961785  tt0118652  alternative              0.0     movie   
64     6961786  tt0118652  alternative              0.0     movie   
65     6961787  tt0118652  alternative              0.0     movie   

            originalTitle  isAdult  startYear runtimeMinutes  \
42  The Attic Expeditions      0.0       2001            100   
43  The Attic Expeditions      0.0       2001            100   
44  The Attic Expeditions      0.0       2001            100   
45  The Attic Expeditions      0.0       2001            100   
46  The Attic Expeditions      0.0       2001            100   
47  The Attic Expeditions      0.0       2001            100   
48  The Attic Expeditions      0.0       2001            100   
49  The Attic Expeditions      0.0       2001            100   
50  The Attic Expeditions      0.0       2001            100   
51  The Attic Expeditions      0.0       2001            100   
52  The Attic Expeditions      0.0       2001            100   
53  The Attic Expeditions      0.0       2001            100   
54  The Attic Expeditions      0.0       2001            100   
55  The Attic Expeditions      0.0       2001            100   
56  The Attic Expeditions      0.0       2001            100   
57  The Attic Expeditions      0.0       2001            100   
58  The Attic Expeditions      0.0       2001            100   
59  The Attic Expeditions      0.0       2001            100   
60  The Attic Expeditions      0.0       2001            100   
61  The Attic Expeditions      0.0       2001            100   
62  The Attic Expeditions      0.0       2001            100   
63  The Attic Expeditions      0.0       2001            100   
64  The Attic Expeditions      0.0       2001            100   
65  The Attic Expeditions      0.0       2001            100   

                   genres  directors  category  averageRating  numVotes  
42  Comedy,Horror,Mystery  nm0440948  composer            4.9    1850.0  
43  Comedy,Horror,Mystery  nm0440948     actor            4.9    1850.0  
44  Comedy,Horror,Mystery  nm0440948   actress            4.9    1850.0  
45  Comedy,Horror,Mystery  nm0440948  director            4.9    1850.0  
46  Comedy,Horror,Mystery  nm0440948    writer            4.9    1850.0  
47  Comedy,Horror,Mystery  nm0440948  producer            4.9    1850.0  
48  Comedy,Horror,Mystery  nm0440948  composer            4.9    1850.0  
49  Comedy,Horror,Mystery  nm0440948     actor            4.9    1850.0  
50  Comedy,

In [12]:
imdb_df[imdb_df['originalTitle'] == 'Akelarre']

Unnamed: 0     tconst        types  isOriginalTitle titleType  \
2795249    32998075  tt9916362  imdbDisplay              0.0     movie   
2795250    32998076  tt9916362  imdbDisplay              0.0     movie   
2795251    32998077  tt9916362  imdbDisplay              0.0     movie   
2795252    32998080  tt9916362  imdbDisplay              0.0     movie   
2795253    32998081  tt9916362  imdbDisplay              0.0     movie   
2795254    32998082  tt9916362  imdbDisplay              0.0     movie   
2795255    32998125  tt9916362     original              1.0     movie   
2795256    32998126  tt9916362     original              1.0     movie   
2795257    32998127  tt9916362     original              1.0     movie   
2795258    32998130  tt9916362     original              1.0     movie   
2795259    32998131  tt9916362     original              1.0     movie   
2795260    32998132  tt9916362     original              1.0     movie   
2795261    32998205  tt9916362  alternative              0.0     movie   
2795262    32998206  tt9916362  alternative              0.0     movie   
2795263    32998207  tt9916362  alternative              0.0     movie   
2795264    32998210  tt9916362  alternative              0.0     movie   
2795265    32998211  tt9916362  alternative              0.0     movie   
2795266    32998212  tt9916362  alternative              0.0     movie   

        originalTitle  isAdult  startYear runtimeMinutes         genres  \
2795249      Akelarre      0.0       2020             92  Drama,History   
2795250      Akelarre      0.0       2020             92  Drama,History   
2795251      Akelarre      0.0       2020             92  Drama,History   
2795252      Akelarre      0.0       2020             92  Drama,History   
2795253      Akelarre      0.0       2020             92  Drama,History   
2795254      Akelarre      0.0       2020             92  Drama,History   
2795255      Akelarre      0.0       2020             92  Drama,History   
2795256      Akelarre      0.0       2020             92  Drama,History   
2795257      Akelarre      0.0       2020             92  Drama,History   
2795258      Akelarre      0.0       2020             92  Drama,History   
2795259      Akelarre      0.0       2020             92  Drama,History   
2795260      Akelarre      0.0       2020             92  Drama,History   
2795261      Akelarre      0.0       2020             92  Drama,History   
2795262      Akelarre      0.0       2020             92  Drama,History   
2795263      Akelarre      0.0       2020             92  Drama,History   
2795264      Akelarre      0.0       2020             92  Drama,History   
2795265      Akelarre      0.0       2020             92  Drama,History   
2795266      Akelarre      0.0       2020             92  Drama,History   

         directors  category  averageRating  numVotes  
2795249  nm1893148  composer            6.4    5179.0  
2795250  nm1893148   actress            6.4    5179.0  
2795251  nm1893148     actor            6.4    5179.0  
2795252  nm1893148  director            6.4    5179.0  
2795253  nm1893148    writer            6.4    5179.0  
2795254  nm1893148  producer            6.4    5179.0  
2795255  nm1893148  composer            6.4    5179.0  
2795256  nm1893148   actress            6.4    5179.0  
2795257  nm1893148     actor            6.4    5179.0  
2795258  nm1893148  director            6.4    5179.0  
2795259  nm1893148    writer            6.4    5179.0  
2795260  nm1893148  producer            6.4    5179.0  
2795261  nm1893148  composer            6.4    5179.0  
2795262  nm1893148   actress            6.4    5179.0  
2795263  nm1893148     actor            6.4    5179.0  
2795264  nm1893148  director            6.4    5179.0  
2795265  nm1893148    writer            6.4    5179.0  
2795266  nm1893148  producer            6.4    5179.0

### Obs.:
Based on the above samples/exploration, I identified a consistent pattern where each media entry has an original type along with variants that do not affect the averageRating or numVotes. It repeats the the category over for each unique value in `types`, which is redundant. I will keep only records for one type in `types` column, as the variants provide no additional value to the analysis. After removing all records and will drop the `types` and `isOriginalTitle` columns as we not longer need them.

In [13]:
# Checking the number of unique titles before the cleaning to make sure we don't lose data
nunique_titles = imdb_df['originalTitle'].nunique()
print("Number of unique titles: ", nunique_titles)

Number of unique titles:  229522


In [14]:
# Removing all observations that are not the original
imdb_original = imdb_df[(imdb_df['types'] == "original") | (imdb_df['types'] == "tv")]

# Checking the number of unique titles after the cleaning to make sure we don't lose data
nunique_imdb_original = imdb_original['originalTitle'].nunique()

# Dataset after trasformation
print("Data dimensions: ", imdb_original.shape)
print("Number of unique titles: ", nunique_imdb_original)
imdb_original.head()

Data dimensions:  (1102454, 14)
Number of unique titles:  207981


Unnamed: 0     tconst     types  isOriginalTitle titleType originalTitle  \
7      6475891  tt0111596  original              1.0     movie         Vazir   
8      6475892  tt0111596  original              1.0     movie         Vazir   
9      6475894  tt0111596  original              1.0     movie         Vazir   
10     6475896  tt0111596  original              1.0     movie         Vazir   
11     6475897  tt0111596  original              1.0     movie         Vazir   

    isAdult  startYear runtimeMinutes              genres   directors  \
7       0.0       2015             \N  Action,Crime,Drama  nm13648797   
8       0.0       2015             \N  Action,Crime,Drama  nm13648797   
9       0.0       2015             \N  Action,Crime,Drama  nm13648797   
10      0.0       2015             \N  Action,Crime,Drama  nm13648797   
11      0.0       2015             \N  Action,Crime,Drama  nm13648797   

    category  averageRating  numVotes  
7     editor            8.1      63.0  
8      actor            8.1      63.0  
9    actress            8.1      63.0  
10  director            8.1      63.0  
11    writer            8.1      63.0

### Obs.:

We did lost some midias that did not have original titles. Let's look at them below.

In [15]:
print("Number of records reduce was: ", imdb_df.shape[0] - imdb_original.shape[0])
print("Number of unique titles lost was: ", nunique_titles - nunique_imdb_original)

Number of records reduce was:  1692841
Number of unique titles lost was:  21541


### Obs.:

Let's go after the lost records.

In [16]:
# List o unique titles after the cleaning
orig_titles = imdb_original['originalTitle'].unique()

# Movies that don't have original title and therefore are not in the the dataset after the cleaning
movies_no_orig = imdb_df[~(imdb_df['originalTitle'].isin(orig_titles))]

# Checking the number of unique titles that are not in the cleaned df
nunique_movies_no_orig = movies_no_orig['originalTitle'].nunique()

# Dataset after trasformation
print("Data dimensions: ", movies_no_orig.shape)
print("Number of unique titles: ", nunique_movies_no_orig)
movies_no_orig.head()

Data dimensions:  (95880, 14)
Number of unique titles:  21541


Unnamed: 0     tconst types  isOriginalTitle titleType  \
409      7702551  tt0144449    \N              0.0     movie   
410      7702552  tt0144449    \N              0.0     movie   
411      7702555  tt0144449    \N              0.0     movie   
412      7702556  tt0144449    \N              0.0     movie   
1011     8274579  tt0176001    \N              0.0     movie   

              originalTitle  isAdult  startYear runtimeMinutes       genres  \
409   Nema aviona za Zagreb      0.0       2012             82    Biography   
410   Nema aviona za Zagreb      0.0       2012             82    Biography   
411   Nema aviona za Zagreb      0.0       2012             82    Biography   
412   Nema aviona za Zagreb      0.0       2012             82    Biography   
1011           Panic Bodies      0.0       2003             70  Documentary   

      directors         category  averageRating  numVotes  
409   nm0309428  cinematographer            8.2       8.0  
410   nm0309428  archive_footage            8.2       8.0  
411   nm0309428           writer            8.2       8.0  
412   nm0309428         producer            8.2       8.0  
1011  nm0393682            actor            7.0      12.0

### Obs.:

The unique values match the 21,541 movies that were indentified as missing after selecting the original titles. Let's recover them.

In [17]:
## Get missing movies
titles_N= movies_no_orig[movies_no_orig['types'] == '\\N']

# Checking the number of unique titles that are not in the cleaned df
nunique_titles_N= titles_N['originalTitle'].nunique()

# Dataset after trasformation
print("Data dimensions: ", titles_N.shape)
print("Number of unique titles: ", nunique_titles_N)
titles_N.head()

Data dimensions:  (95332, 14)
Number of unique titles:  21507


Unnamed: 0     tconst types  isOriginalTitle titleType  \
409      7702551  tt0144449    \N              0.0     movie   
410      7702552  tt0144449    \N              0.0     movie   
411      7702555  tt0144449    \N              0.0     movie   
412      7702556  tt0144449    \N              0.0     movie   
1011     8274579  tt0176001    \N              0.0     movie   

              originalTitle  isAdult  startYear runtimeMinutes       genres  \
409   Nema aviona za Zagreb      0.0       2012             82    Biography   
410   Nema aviona za Zagreb      0.0       2012             82    Biography   
411   Nema aviona za Zagreb      0.0       2012             82    Biography   
412   Nema aviona za Zagreb      0.0       2012             82    Biography   
1011           Panic Bodies      0.0       2003             70  Documentary   

      directors         category  averageRating  numVotes  
409   nm0309428  cinematographer            8.2       8.0  
410   nm0309428  archive_footage            8.2       8.0  
411   nm0309428           writer            8.2       8.0  
412   nm0309428         producer            8.2       8.0  
1011  nm0393682            actor            7.0      12.0

In [18]:
print("Still missing {} movies" .format(nunique_movies_no_orig - nunique_titles_N))

Still missing 34 movies


In [19]:
movies_no_orig['types'].unique()

array(['\\N', 'imdbDisplay', 'working', 'dvd', 'festival', 'alternative',
       'video'], dtype=object)

In [20]:
# List o unique titles after removing original and nulls
last_list = titles_N['originalTitle'].unique()

# Movies that are missing
movies_no_orig_null = movies_no_orig[~(movies_no_orig['originalTitle'].isin(last_list))]

# Checking the number of unique titles that are not in the cleaned df
nunique_missing = movies_no_orig_null['originalTitle'].nunique()

# Display
print("Number of unique titles: ", nunique_missing)

Number of unique titles:  34


In [21]:
movies_no_orig_null['types'].unique()

array(['imdbDisplay', 'working', 'alternative', 'festival'], dtype=object)

In [22]:
# Get last 34 movies
titles_imdb = movies_no_orig_null[movies_no_orig_null['types'] == 'imdbDisplay']

# Confirm the unique count
nunique_titles_imdb = titles_imdb['originalTitle'].nunique()

# Dataset after trasformation
print("Data dimensions: ", titles_imdb.shape)
print("Number of unique titles: ", nunique_titles_imdb)
titles_imdb.head()

Data dimensions:  (161, 14)
Number of unique titles:  34


Unnamed: 0     tconst        types  isOriginalTitle titleType  \
25495    10485760  tt0315056  imdbDisplay              0.0  tvSeries   
59030    11239400  tt0371599  imdbDisplay              0.0     movie   
59031    11239401  tt0371599  imdbDisplay              0.0     movie   
59032    11239403  tt0371599  imdbDisplay              0.0     movie   
59033    11239405  tt0371599  imdbDisplay              0.0     movie   

                   originalTitle  isAdult  startYear runtimeMinutes  \
25495                 Saia Justa      0.0       2002             60   
59030  Ein Goldfisch unter Haien      0.0       2004             90   
59031  Ein Goldfisch unter Haien      0.0       2004             90   
59032  Ein Goldfisch unter Haien      0.0       2004             90   
59033  Ein Goldfisch unter Haien      0.0       2004             90   

               genres                       directors  category  \
25495       Talk-Show  nm11597242,nm1381005,nm1166794      self   
59030  Comedy,Romance                       nm0090596  composer   
59031  Comedy,Romance                       nm0090596   actress   
59032  Comedy,Romance                       nm0090596     actor   
59033  Comedy,Romance                       nm0090596  director   

       averageRating  numVotes  
25495            6.7      29.0  
59030            6.7      32.0  
59031            6.7      32.0  
59032            6.7      32.0  
59033            6.7      32.0

### Obs.:
It got a little confusing, but at this point, we can significantly reduce the volume and redundancy of the data without loosing any significant information. Let's now consolidate the dasets and get all 229,522 initial count of unique titles.

In [46]:
# Concatenanting all datasets
imdb_clean = pd.concat([imdb_original, titles_N, titles_imdb], ignore_index=True)

# Removing unecessary columns
imdb_clean = imdb_clean.drop(columns=['Unnamed: 0', 'types', 'isOriginalTitle'])

# Confirm the unique count
nunique_imdb_clean = imdb_clean['originalTitle'].nunique()

# Dataset after trasformation
print("Data dimensions: ", imdb_clean.shape)
print("Number of unique titles: ", nunique_imdb_clean)
imdb_clean.head()

Data dimensions:  (1197947, 11)
Number of unique titles:  229522


tconst titleType originalTitle  isAdult  startYear runtimeMinutes  \
0  tt0111596     movie         Vazir      0.0       2015             \N   
1  tt0111596     movie         Vazir      0.0       2015             \N   
2  tt0111596     movie         Vazir      0.0       2015             \N   
3  tt0111596     movie         Vazir      0.0       2015             \N   
4  tt0111596     movie         Vazir      0.0       2015             \N   

               genres   directors  category  averageRating  numVotes  
0  Action,Crime,Drama  nm13648797    editor            8.1      63.0  
1  Action,Crime,Drama  nm13648797     actor            8.1      63.0  
2  Action,Crime,Drama  nm13648797   actress            8.1      63.0  
3  Action,Crime,Drama  nm13648797  director            8.1      63.0  
4  Action,Crime,Drama  nm13648797    writer            8.1      63.0

In [47]:
print("Number of records reduced was: ", imdb_df.shape[0] - imdb_clean.shape[0])
print("Number of unique titles lost was: ", nunique_titles - nunique_imdb_clean)

Number of records reduced was:  1597348
Number of unique titles lost was:  0


### Obs.:
From the above line, we can see the number of records was significantly reduced without loosing any unique title. \

Moving forward, I still see an opportunity to reduce redundancy further in the dataset. I believe we can have one record per midia if we one-hot encode the `category` column.

In [48]:
print("We will have {} more columns in the dataset." .format(imdb_clean['category'].nunique() - 2))

We will have 10 more columns in the dataset.


In [49]:
# One-hot encode the 'category' column
imdb_one_hot = pd.get_dummies(imdb_clean, columns=['category'], drop_first=True, dtype=int)

# Group by movie identifiers and aggregate numeric columns
imdb_grouped = imdb_one_hot.groupby(
    ['tconst', 'titleType', 'originalTitle', 'isAdult', 'startYear', 'runtimeMinutes', 'genres', 'directors'],
    as_index=False
).agg({
    'averageRating': 'mean',
    'numVotes': 'mean',     
    **{col: 'sum' for col in imdb_one_hot.columns if col.startswith(('category_'))}
})

# Remove the "category_" prefix from the one-hot encoded column names
imdb_grouped.columns = imdb_grouped.columns.str.replace('category_', '')

# Dataset after transformation
print("Data dimensions: ", imdb_grouped.shape)
print("Number of unique titles: ", imdb_grouped['originalTitle'].nunique())
imdb_grouped.head()


Data dimensions:  (252763, 21)
Number of unique titles:  229522


tconst titleType          originalTitle  isAdult  startYear  \
0  tt0111596     movie                  Vazir      0.0       2015   
1  tt0118589     movie                Glitter      0.0       2001   
2  tt0118652     movie  The Attic Expeditions      0.0       2001   
3  tt0121164     movie           Corpse Bride      0.0       2005   
4  tt0121389     short          I, an Actress      0.0       2005   

  runtimeMinutes                  genres            directors  averageRating  \
0             \N      Action,Crime,Drama           nm13648797            8.1   
1            104     Drama,Music,Romance            nm0193554            2.4   
2            100   Comedy,Horror,Mystery            nm0440948            4.9   
3             77  Animation,Drama,Family  nm0000318,nm0425843            7.3   
4             10                   Short            nm0473647            6.0   

   numVotes  actress  archive_footage  archive_sound  cinematographer  \
0      63.0        1                0              0                1   
1   23686.0        1                0              0                1   
2    1850.0        1                0              0                0   
3  279860.0        1                0              0                0   
4     343.0        1                0              0                1   

   composer  director  editor  producer  production_designer  self  writer  
0         1         1       1         0                    0     0       1  
1         1         1       0         1                    0     0       1  
2         1         1       0         1                    0     0       1  
3         0         1       0         0                    0     0       1  
4         0         0       0         0                    0     0       0

In [50]:
imdb_grouped[imdb_grouped['originalTitle'].duplicated()]

tconst  titleType       originalTitle  isAdult  startYear  \
986     tt0292190  videoGame  Planet of the Apes      0.0       2001   
1013    tt0292508      movie               Fidel      0.0       2001   
1300    tt0303354      movie        Killing Time      0.0       2002   
1412    tt0303938      movie                Dust      0.0       2001   
1521    tt0315355      movie         The Calling      0.0       2002   
...           ...        ...                 ...      ...        ...   
252737  tt9913936      movie            Paradise      0.0       2019   
252742  tt9914642      movie           Albatross      0.0       2017   
252748  tt9915138   tvSeries          Scandalous      0.0       2018   
252753  tt9916038      short                 Eco      0.0       2019   
252760  tt9916460    tvMovie           Pink Taxi      0.0       2019   

       runtimeMinutes                   genres   directors  averageRating  \
986                \N  Action,Adventure,Sci-Fi          \N            5.3   
1013               91    Biography,Documentary   nm0106073            7.5   
1300               81                       \N   nm1097510            5.9   
1412               90   Comedy,Horror,Thriller   nm0556573            4.6   
1521              117             Drama,Family   nm1161445            7.2   
...               ...                      ...         ...            ...   
252737            135              Crime,Drama  nm10091596            7.5   
252742             97              Documentary   nm5300859            8.3   
252748             60        Crime,Documentary   nm7710846            6.6   
252753             14              Drama,Short   nm1517212            6.4   
252760             \N                   Comedy   nm7048843            9.4   

        numVotes  actress  archive_footage  archive_sound  cinematographer  \
986        100.0        1                0              0                0   
1013       558.0        0                0              0                2   
1300        60.0        1                0              0                0   
1412        98.0        1                0              0                1   
1521        13.0        1                0              0                1   
...          ...      ...              ...            ...              ...   
252737      59.0        0                0              0                1   
252742      42.0        0                0              0                1   
252748      11.0        0                0              0                0   
252753      15.0        1                0              0                1   
252760      18.0        1                0              0                1   

        composer  director  editor  producer  production_designer  self  \
986            1         0       0         0                    0     0   
1013           2         2       0         2                    0     2   
1300           1         1       1         1                    1     0   
1412           1         1       1         0                    0     0   
1521           1         1       0         1                    0     0   
...          ...       ...     ...       ...                  ...   ...   
252737         0         0       0         0                    0     0   
252742         0         1       0         0                    0     0   
252748         0         0       0         0                    0     1   
252753         1         1       0         1                    0     0   
252760         0         1       0         0                    0     0   

        writer  
986          1  
1013         0  
1300         0  
1412         0  
1521         0  
...        ...  
252737       0  
252742       1  
252748       0  
252753       1  
252760       0  

[23241 rows x 21 columns]

In [51]:
imdb_grouped[imdb_grouped['originalTitle'] == 'Planet of the Apes']

tconst  titleType       originalTitle  isAdult  startYear  \
10   tt0133152      movie  Planet of the Apes      0.0       2001   
986  tt0292190  videoGame  Planet of the Apes      0.0       2001   

    runtimeMinutes                   genres  directors  averageRating  \
10             119  Action,Adventure,Sci-Fi  nm0000318            5.7   
986             \N  Action,Adventure,Sci-Fi         \N            5.3   

     numVotes  actress  archive_footage  archive_sound  cinematographer  \
10   221998.0        1                0              0                0   
986     100.0        1                0              0                0   

     composer  director  editor  producer  production_designer  self  writer  
10          0         1       0         1                    0     0       1  
986         1         0       0         0                    0     0       1

### Obs.:

Now that we got to the point of one record per movie, we can explore ways of working with `genres`, `directors`, and `titleType`, which are categorical columns.

### `genres`

Here, I'll create a numeric mapping for genres and split the genres into up to three columns (genre1, genre2, genre3), each representing a unique numeric value for a genre. If a movie had less than three genres, the remaining columns were filled with 0. If it has more, I will just use the first 3.

In [52]:
# Split the genres string by commas and explode the resulting list into individual rows
genres_split = imdb_grouped['genres'].str.split(',')

# Flatten the list of genres and get unique values
unique_genres = pd.Series([genre.strip() for sublist in genres_split for genre in sublist]).unique()

# Display the unique genres
print(unique_genres)

['Action' 'Crime' 'Drama' 'Music' 'Romance' 'Comedy' 'Horror' 'Mystery'
 'Animation' 'Family' 'Short' 'Adventure' 'Fantasy' 'Thriller' 'Sci-Fi'
 'Documentary' '\\N' 'Biography' 'Sport' 'History' 'War' 'Western'
 'Musical' 'Talk-Show' 'Adult' 'Reality-TV' 'Game-Show' 'News']


In [53]:
# Create a mapping of genres to numeric values
genre_mapping = {
    'Null': 0, 'Action': 1, 'Crime': 2, 'Drama': 3, 'Music': 4, 'Romance': 5, 'Comedy': 6,
    'Horror': 7, 'Mystery': 8, 'Animation': 9, 'Family': 10, 'Short': 11, 'Adventure': 12,
    'Fantasy': 13, 'Thriller': 14, 'Sci-Fi': 15, 'Documentary': 16, 'Biography': 17, 'Sport': 18,
    'History': 19, 'War': 20, 'Western': 21, 'Musical': 22, 'Talk-Show': 23, 'Adult': 24,
    'Reality-TV': 25, 'Game-Show': 26, 'News': 27, '\\N': 0
}

# Create a function to split genres and assign numeric values to new columns
def assign_genres(genres_list):
    # Split the genres and map to numbers
    mapped_genres = [genre_mapping[genre.strip()] for genre in genres_list if genre in genre_mapping]
    
    # Assign up to 3 genres (fill with 0 if there are fewer than 3 genres)
    return mapped_genres[:3] + [0] * (3 - len(mapped_genres))

# Apply the function to create new genre columns
imdb_grouped[['genre1', 'genre2', 'genre3']] = imdb_grouped['genres'].str.split(',').apply(assign_genres).apply(pd.Series)

# Display the transformed data
imdb_grouped[['genres', 'genre1', 'genre2', 'genre3']]

genres  genre1  genre2  genre3
0           Action,Crime,Drama       1       2       3
1          Drama,Music,Romance       3       4       5
2        Comedy,Horror,Mystery       6       7       8
3       Animation,Drama,Family       9       3      10
4                        Short      11       0       0
...                        ...     ...     ...     ...
252758                Thriller      14       0       0
252759           Drama,History       3      19       0
252760                  Comedy       6       0       0
252761                   Drama       3       0       0
252762                      \N       0       0       0

[252763 rows x 4 columns]

### `directors`

Here, I will split the directors column by removing the 'nm' prefix and create two columns (director1, director2). If only one director was listed, the second column was set to 0. Missing values ('\\N') were handled by assigning 0.

In [54]:
# Create a function to process the directors column
def assign_directors(directors_list):
    # Remove 'nm' prefix and split by commas
    processed_directors = [int(director.strip().replace('nm', '')) if director != '\\N' else 0 for director in directors_list.split(',')]
    
    # Assign up to 2 directors and fill with 0 if there is only one director
    return processed_directors[:2] + [0] * (2 - len(processed_directors))

# Apply the function to create new director columns
imdb_grouped[['director1', 'director2']] = imdb_grouped['directors'].apply(assign_directors).apply(pd.Series).astype(int)

# Display the transformed data
imdb_grouped[['directors', 'director1', 'director2']].head()


directors  director1  director2
0           nm13648797   13648797          0
1            nm0193554     193554          0
2            nm0440948     440948          0
3  nm0000318,nm0425843        318     425843
4            nm0473647     473647          0

### `titleType`

Here, I will map each unique value in the titleType column to a corresponding numeric value using a dictionary, and create a new column (titleType_encoded) for this numeric column.

In [55]:
imdb_grouped['titleType'].unique()

array(['movie', 'short', 'video', 'tvMovie', 'tvMiniSeries', 'tvSeries',
       'videoGame', 'tvSpecial', 'tvShort', 'tvEpisode'], dtype=object)

In [56]:
# Create a mapping of titleType to numeric values
titleType_mapping = {
    'movie': 1,
    'short': 2,
    'video': 3,
    'tvMovie': 4,
    'tvMiniSeries': 5,
    'tvSeries': 6,
    'videoGame': 7,
    'tvSpecial': 8,
    'tvShort': 9,
    'tvEpisode': 10
}

# Apply the mapping to the 'titleType' column
imdb_grouped['titleType_encoded'] = imdb_grouped['titleType'].map(titleType_mapping)

# Display the transformed data
imdb_grouped[['titleType', 'titleType_encoded']].head()


titleType  titleType_encoded
0     movie                  1
1     movie                  1
2     movie                  1
3     movie                  1
4     short                  2

### Find and fill null values

### `runtimeMinutes`

Here, I will fill null values in `runtimeMinutes` with the average runtime per title in `titleType` column.

In [57]:
imdb_grouped['runtimeMinutes'].value_counts()

runtimeMinutes
\N     43683
90      9680
60      6489
30      5156
15      4832
       ...  
311        1
413        1
624        1
529        1
462        1
Name: count, Length: 546, dtype: int64

In [58]:
# Get average runtime for each title type
imdb_grouped['runtimeMinutes'] = np.where(imdb_grouped['runtimeMinutes'] == '\\N', np.nan, imdb_grouped['runtimeMinutes']).astype(float)
avg_runtime = imdb_grouped.dropna()
avg_runtime = avg_runtime.groupby(['titleType']).agg(runtimeMinutes_avg=('runtimeMinutes', 'mean')).reset_index()
avg_runtime.head()

titleType  runtimeMinutes_avg
0         movie           95.903562
1         short           15.353863
2     tvEpisode           68.024096
3  tvMiniSeries          105.786276
4       tvMovie           80.757631

In [59]:
# Filling Null values
imdb_merged = imdb_grouped.merge(avg_runtime, on=['titleType'], how='left')
imdb_merged['runtimeMinutes'] = np.where(imdb_merged['runtimeMinutes'].isna(), imdb_merged['runtimeMinutes_avg'], imdb_merged['runtimeMinutes'])
imdb_merged[imdb_merged['runtimeMinutes'].isna()]

Empty DataFrame
Columns: [tconst, titleType, originalTitle, isAdult, startYear, runtimeMinutes, genres, directors, averageRating, numVotes, actress, archive_footage, archive_sound, cinematographer, composer, director, editor, producer, production_designer, self, writer, genre1, genre2, genre3, director1, director2, titleType_encoded, runtimeMinutes_avg]
Index: []

In [60]:
imdb_merged

tconst titleType             originalTitle  isAdult  startYear  \
0       tt0111596     movie                     Vazir      0.0       2015   
1       tt0118589     movie                   Glitter      0.0       2001   
2       tt0118652     movie     The Attic Expeditions      0.0       2001   
3       tt0121164     movie              Corpse Bride      0.0       2005   
4       tt0121389     short             I, an Actress      0.0       2005   
...           ...       ...                       ...      ...        ...   
252758  tt9916270     movie  Il talento del calabrone      0.0       2020   
252759  tt9916362     movie                  Akelarre      0.0       2020   
252760  tt9916460   tvMovie                 Pink Taxi      0.0       2019   
252761  tt9916538     movie       Kuambil Lagi Hatiku      0.0       2019   
252762  tt9916730     movie                    6 Gunn      0.0       2017   

        runtimeMinutes                  genres            directors  \
0            95.903562      Action,Crime,Drama           nm13648797   
1           104.000000     Drama,Music,Romance            nm0193554   
2           100.000000   Comedy,Horror,Mystery            nm0440948   
3            77.000000  Animation,Drama,Family  nm0000318,nm0425843   
4            10.000000                   Short            nm0473647   
...                ...                     ...                  ...   
252758       84.000000                Thriller            nm1480867   
252759       92.000000           Drama,History            nm1893148   
252760       80.757631                  Comedy            nm7048843   
252761      123.000000                   Drama            nm4457074   
252762      116.000000                      \N           nm10538612   

        averageRating  numVotes  actress  archive_footage  archive_sound  \
0                 8.1      63.0        1                0              0   
1                 2.4   23686.0        1                0              0   
2                 4.9    1850.0        1                0              0   
3                 7.3  279860.0        1                0              0   
4                 6.0     343.0        1                0              0   
...               ...       ...      ...              ...            ...   
252758            5.8    1393.0        1                0              0   
252759            6.4    5179.0        1                0              0   
252760            9.4      18.0        1                0              0   
252761            8.3       6.0        1                0              0   
252762            8.1       9.0        0                0              0   

        cinematographer  composer  director  editor  producer  \
0                     1         1         1       1         0   
1                     1         1         1       0         1   
2                     0         1         1       0         1   
3                     0         0         1       0         0   
4                     1         0         0       0         0   
...                 ...       ...       ...     ...       ...   
252758                1         1         1       0         1   
252759                0         1         1       0         1   
252760                1         0         1       0         0   
252761                0         1         1       0         1   
252762                1         0         1       1         1   

        production_designer  self  writer  genre1  genre2  genre3  director1  \
0                         0     0       1       1       2       3   13648797   
1                         0     0       1       3       4       5     193554   
2                         0     0       1       6       7       8     440948   
3                         0     0       1       9       3      10        318   
4                         0     0       0      11       0       0     473647   
...                     ...   ...     ...     ...     ...  

### `numVotes`

No null values found.

In [61]:
imdb_merged[imdb_merged['numVotes'] == '\\N']

Empty DataFrame
Columns: [tconst, titleType, originalTitle, isAdult, startYear, runtimeMinutes, genres, directors, averageRating, numVotes, actress, archive_footage, archive_sound, cinematographer, composer, director, editor, producer, production_designer, self, writer, genre1, genre2, genre3, director1, director2, titleType_encoded, runtimeMinutes_avg]
Index: []

In [62]:
imdb_merged[imdb_merged['numVotes'].isna()]

Empty DataFrame
Columns: [tconst, titleType, originalTitle, isAdult, startYear, runtimeMinutes, genres, directors, averageRating, numVotes, actress, archive_footage, archive_sound, cinematographer, composer, director, editor, producer, production_designer, self, writer, genre1, genre2, genre3, director1, director2, titleType_encoded, runtimeMinutes_avg]
Index: []

### `startYear`
No null values found.

In [63]:
imdb_merged[imdb_merged['startYear'].isna()]

Empty DataFrame
Columns: [tconst, titleType, originalTitle, isAdult, startYear, runtimeMinutes, genres, directors, averageRating, numVotes, actress, archive_footage, archive_sound, cinematographer, composer, director, editor, producer, production_designer, self, writer, genre1, genre2, genre3, director1, director2, titleType_encoded, runtimeMinutes_avg]
Index: []

In [64]:
imdb_merged['startYear'].value_counts()

startYear
2016    16411
2017    15878
2015    15711
2013    15323
2014    15120
2012    15035
2011    14496
2018    13794
2010    13454
2019    13097
2021    12870
2009    12868
2020    12750
2008    11462
2022    11308
2007    10088
2006     9558
2005     7534
2004     5409
2003     4295
2002     3075
2001     2457
2023      770
Name: count, dtype: int64

### Create `rating` column derived from `averageRating`

To classify movies into 5 star-based rating categories, I applied quantile-based binning to the averageRating column since averageRating is a continuous variable amd cannot be used directly in a classification model. This approach divides the data into five groups of similar size, with each bin representing about 20% of the ratings. By using quantiles, the process ensures that each category contains a similar number of movies, resulting in a balanced distribution. This approach may improve the model’s performance by preventing bias toward overrepresented ratings and makes the data suitable for classification.

In [65]:
# Analyze the distribution of averageRating
imdb_merged['averageRating'].describe()

count    252763.000000
mean          6.652431
std           1.518086
min           1.000000
25%           5.800000
50%           6.800000
75%           7.700000
max          10.000000
Name: averageRating, dtype: float64

In [72]:
# Create a new rating column using quantile-based binning
imdb_merged['rating'], bin_edges = pd.qcut(imdb_merged['averageRating'], q=5, labels=[1, 2, 3, 4, 5], retbins=True)
imdb_merged['rating'] = imdb_merged['rating'].astype(int)

# Display the bin edges (ranges)
print("Stars average ratings range: ")
for i in range(1, len(bin_edges)):
    print(f"{i} star: {bin_edges[i-1]:.2f} to {bin_edges[i]:.2f}")

print("")

# Check the distribution of the new rating column
print("New rating distribution: ")
print(imdb_merged['rating'].value_counts())

Stars average ratings range: 
1 star: 1.00 to 5.50
2 star: 5.50 to 6.40
3 star: 6.40 to 7.10
4 star: 7.10 to 7.90
5 star: 7.90 to 10.00

New rating distribution: 
rating
1    53403
4    50778
5    50101
3    49547
2    48934
Name: count, dtype: int64


### Removing unecessary columns

In [73]:
imdb_merged.head()

tconst titleType          originalTitle  isAdult  startYear  \
0  tt0111596     movie                  Vazir      0.0       2015   
1  tt0118589     movie                Glitter      0.0       2001   
2  tt0118652     movie  The Attic Expeditions      0.0       2001   
3  tt0121164     movie           Corpse Bride      0.0       2005   
4  tt0121389     short          I, an Actress      0.0       2005   

   runtimeMinutes                  genres            directors  averageRating  \
0       95.903562      Action,Crime,Drama           nm13648797            8.1   
1      104.000000     Drama,Music,Romance            nm0193554            2.4   
2      100.000000   Comedy,Horror,Mystery            nm0440948            4.9   
3       77.000000  Animation,Drama,Family  nm0000318,nm0425843            7.3   
4       10.000000                   Short            nm0473647            6.0   

   numVotes  actress  archive_footage  archive_sound  cinematographer  \
0      63.0        1                0              0                1   
1   23686.0        1                0              0                1   
2    1850.0        1                0              0                0   
3  279860.0        1                0              0                0   
4     343.0        1                0              0                1   

   composer  director  editor  producer  production_designer  self  writer  \
0         1         1       1         0                    0     0       1   
1         1         1       0         1                    0     0       1   
2         1         1       0         1                    0     0       1   
3         0         1       0         0                    0     0       1   
4         0         0       0         0                    0     0       0   

   genre1  genre2  genre3  director1  director2  titleType_encoded  \
0       1       2       3   13648797          0                  1   
1       3       4       5     193554          0                  1   
2       6       7       8     440948          0                  1   
3       9       3      10        318     425843                  1   
4      11       0       0     473647          0                  2   

   runtimeMinutes_avg  rating  
0           95.903562       5  
1           95.903562       1  
2           95.903562       1  
3           95.903562       4  
4           15.353863       2

In [74]:
# Removing unecessary columns
imdb_final = imdb_merged.drop(columns=['tconst', 'originalTitle', 'titleType', 'genres', 'directors', 'runtimeMinutes_avg'])
imdb_final

isAdult  startYear  runtimeMinutes  averageRating  numVotes  actress  \
0           0.0       2015       95.903562            8.1      63.0        1   
1           0.0       2001      104.000000            2.4   23686.0        1   
2           0.0       2001      100.000000            4.9    1850.0        1   
3           0.0       2005       77.000000            7.3  279860.0        1   
4           0.0       2005       10.000000            6.0     343.0        1   
...         ...        ...             ...            ...       ...      ...   
252758      0.0       2020       84.000000            5.8    1393.0        1   
252759      0.0       2020       92.000000            6.4    5179.0        1   
252760      0.0       2019       80.757631            9.4      18.0        1   
252761      0.0       2019      123.000000            8.3       6.0        1   
252762      0.0       2017      116.000000            8.1       9.0        0   

        archive_footage  archive_sound  cinematographer  composer  director  \
0                     0              0                1         1         1   
1                     0              0                1         1         1   
2                     0              0                0         1         1   
3                     0              0                0         0         1   
4                     0              0                1         0         0   
...                 ...            ...              ...       ...       ...   
252758                0              0                1         1         1   
252759                0              0                0         1         1   
252760                0              0                1         0         1   
252761                0              0                0         1         1   
252762                0              0                1         0         1   

        editor  producer  production_designer  self  writer  genre1  genre2  \
0            1         0                    0     0       1       1       2   
1            0         1                    0     0       1       3       4   
2            0         1                    0     0       1       6       7   
3            0         0                    0     0       1       9       3   
4            0         0                    0     0       0      11       0   
...        ...       ...                  ...   ...     ...     ...     ...   
252758       0         1                    0     0       1      14       0   
252759       0         1                    0     0       1       3      19   
252760       0         0                    0     0       0       6       0   
252761       0         1                    0     0       1       3       0   
252762       1         1                    0     0       0       0       0   

        genre3  director1  director2  titleType_encoded  rating  
0            3   13648797          0                  1       5  
1            5     193554          0                  1       1  
2            8     440948          0                  1       1  
3           10        318     425843                  1       4  
4            0     473647          0                  2       2  
...        ...        ...        ...                ...     ...  
252758       0    1480867          0                  1       2  
252759       0    1893148          0                  1       2  
252760       0    7048843          0                  4       5  
252761       0    4457074          0                  1       5  
252762       0   10538612          0                  1       5  

[252763 rows x 23 columns]

In [75]:
imdb_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252763 entries, 0 to 252762
Data columns (total 23 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   isAdult              252763 non-null  float64
 1   startYear            252763 non-null  int64  
 2   runtimeMinutes       252763 non-null  float64
 3   averageRating        252763 non-null  float64
 4   numVotes             252763 non-null  float64
 5   actress              252763 non-null  int64  
 6   archive_footage      252763 non-null  int64  
 7   archive_sound        252763 non-null  int64  
 8   cinematographer      252763 non-null  int64  
 9   composer             252763 non-null  int64  
 10  director             252763 non-null  int64  
 11  editor               252763 non-null  int64  
 12  producer             252763 non-null  int64  
 13  production_designer  252763 non-null  int64  
 14  self                 252763 non-null  int64  
 15  writer           

### Export clean data

In [76]:
file_name = "newimdb_clean.csv"

imdb_final.to_csv(data_loc + file_name, index=False)

* Begin the data cleaning process by addressing duplicate entries in the dataset, with a particular focus on the 'category' and 'directors' columns. Remove any redundant rows and determine the best approach for handling variations within these columns. You may choose to either aggregate the diverse values into lists or select the most frequent entry for each instance; explain the reasoning behind your chosen method. Subsequently, identify and eliminate irrelevant columns, such as 'Unnamed: 0', which do not contribute meaningful information for regression or classification tasks. Provide a clear justification for the removal of each column, taking into account their potential significance in model development. Your explanation should demonstrate a thorough understanding of the dataset's structure and the relevance of each feature to the intended analysis.

* Analyze the dataset for missing values, particularly in columns like 'runtimeMinutes' and 'genres'. Determine whether to remove rows with missing data or impute values using appropriate statistical methods (mean, median, or mode), and justify your approach. For feature engineering, transform non-numerical columns such as 'genres' and 'directors' using one-hot encoding or label encoding techniques. Consider grouping genres into broader categories to potentially improve model performance. Provide clear explanations for your strategies in handling missing values and implementing feature engineering, ensuring your decisions are data-driven and aligned with the goals of the analysis.
  
* Conduct exploratory data analysis (EDA) on all remaining columns. Handle non-numerical values appropriately, convert data types as needed, and apply transformations or standardization where necessary. Implement dummy coding for categorical variables. For classification purposes, create a new 'Rating' column derived from 'averageRating'. Instead of using a predetermined scale, devise a binning strategy that results in a balanced class distribution across 5 rating categories (1 to 5 stars). Explain your chosen binning method and analyze the resulting distribution. Discuss how this balanced approach may impact your classification model's performance compared to using the original 'averageRating' values.

### Problem 2.  
You will make regression models here using the cleaned data from problem 1. Use averageRating as the target variable for this problem.
* Begin by selecting five advanced machine learning models for multiple linear/nonlinear regression, excluding basic models like OLS, Decision Tree, Random Forest, K-Nearest Neighbors, and SVM. Determine appropriate test size and validation method, providing a clear rationale for your choices. Split your data into training and testing sets based on these decisions. Use 'averageRating' as your target variable and preprocess your feature set as required by your selected models.
  
* Implement each of your five chosen regression models using the prepared data. Train each model on the training set and generate predictions for the test set. Create a dataframe that combines the predictions from all five models for the test data, and display the first few rows of this dataframe.
  
* Evaluate the performance of all five models by calculating the Root Mean Square Error (RMSE) and R-squared values for both the training and testing data. Present these metrics in a clear, tabular format for easy comparison. Based on these results, select the best-performing model and provide a detailed explanation for your choice. Finally, critically assess whether your best model is suitable for production use. Discuss its strengths and limitations, and propose specific strategies or techniques you would employ to further improve the model's performance.

In [77]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)

In [79]:
data_loc = "/Users/gabrielvictorgomesferreira/Library/Mobile Documents/com~apple~CloudDocs/Work/ISU Classes/MIS 546 - Advanced Business Analytics/Data/"
file_name = "newimdb_clean.csv"
clean_df = pd.read_csv(data_loc + file_name)
clean_df.head()

isAdult  startYear  runtimeMinutes  averageRating  numVotes  actress  \
0      0.0       2015       95.903562            8.1      63.0        1   
1      0.0       2001      104.000000            2.4   23686.0        1   
2      0.0       2001      100.000000            4.9    1850.0        1   
3      0.0       2005       77.000000            7.3  279860.0        1   
4      0.0       2005       10.000000            6.0     343.0        1   

   archive_footage  archive_sound  cinematographer  composer  director  \
0                0              0                1         1         1   
1                0              0                1         1         1   
2                0              0                0         1         1   
3                0              0                0         0         1   
4                0              0                1         0         0   

   editor  producer  production_designer  self  writer  genre1  genre2  \
0       1         0                    0     0       1       1       2   
1       0         1                    0     0       1       3       4   
2       0         1                    0     0       1       6       7   
3       0         0                    0     0       1       9       3   
4       0         0                    0     0       0      11       0   

   genre3  director1  director2  titleType_encoded  rating  
0       3   13648797          0                  1       5  
1       5     193554          0                  1       1  
2       8     440948          0                  1       1  
3      10        318     425843                  1       4  
4       0     473647          0                  2       2

### Problem 3.  
You will make classification models here. Use Rating as the target variable. Make sure to exclude averageRating(Why?) from the predictor variables.

* Select five advanced classification models beyond basic ones like Multinomial Logistic Regression, Decision Tree, Random Forest, K-Nearest Neighbors, and SVM. Determine appropriate test size and validation method, providing a clear rationale for your choices. Split your data into training and testing sets based on these decisions. Use 'Rating' as your target variable. Explain why 'averageRating' should be excluded. Preprocess your feature set as required by your selected models.

* Implement each of your five chosen classification models using the prepared data. Train each model on the training set and generate predictions for the test set. Create a table showing the accuracy scores for all five models on both the training and testing data. For your best-performing model, create and display a colored heatmap of the confusion matrix for the test data predictions.

* Select one additional metric beyond accuracy for validating your models, explaining why you chose this metric. Evaluate all five models using this additional metric and determine which model performs best based on it. Critically assess whether your best model is suitable for production use, discussing its strengths and limitations. Propose specific strategies or techniques you would employ to further improve the model's performance.

### Problem 4. 
For this problem, you'll create clustering models using techniques other than K-Means clustering.
Here's the clustering problem structured into three parts:

* Select relevant columns from the cleaned data for clustering, providing a brief explanation for your choices. Determine the appropriate number of clusters (K) you want to use and justify your reasoning. This step should involve careful consideration of the dataset's characteristics and the potential insights you aim to gain from the clustering process.

* Implement two clustering models using techniques other than K-Means clustering with the chosen features. Predict cluster assignments for data points using each model. Add these cluster labels to the original dataset from problem 1, creating new columns to represent the cluster assignments from each model.

* For each of the two clustering models, identify two new pieces of information or insights about the dataset using the added cluster labels. Analyze how the clusters relate to other variables in the dataset and what patterns or groupings they reveal. Explain your process for deriving these insights and discuss their potential implications for understanding the IMDb dataset. Consider how these insights might be valuable for further analysis or decision-making in the context of movie ratings and characteristics.

### Problem 5. 
Upload the AirPassenger data. The dataset has two columns: Month representing the time in a specific format (e.g., "1949-01" for January 1949) and #Passengers representing the number of passengers for each corresponding month.

*  Extract month and year from the 'Month' column. Perform summary statistics on the '#Passengers' column. Visualize the overall distribution of passenger counts using appropriate plots. Create time series plots to visually inspect trends, seasonality, and patterns. Decompose the time series into trend, seasonality, and residual components, and visualize each component separately.

* Apply ARIMA, Exponential Smoothing, and Prophet models to the data. Evaluate the accuracy of all three forecasting methods using appropriate metrics such as RMSE, MAE, and MAPE. Create autocorrelation and partial autocorrelation plots to identify lag relationships in the time series data. Implement and visualize a technique like moving averages or differencing to remove trends and highlight patterns in the data.

* Implement an anomaly detection method, such as a rolling average with anomaly thresholds. Visualize and analyze the identified anomalies in the context of the time series. Discuss the implications of these anomalies and how they relate to the overall patterns observed in the data. Compare the results of the anomaly detection with the findings from the forecasting models, and explain any discrepancies or correlations between the two analyses.
